docs: app integrity documentation (#235)
govuk-one-login/mobile-ios-one-login-app: AppIntegrity/README.md

#  App Integrity

## Package description

This package contains the logic to implement app integrity checks using [Firebase App Check](https://firebase.google.com/docs/app-check). It therefore encorporates the Firebase SDK as a dependency, along with the Networking module to enable backend requests.

## Overview

### What is app integrity?

App Integrity describes checks made (in this case, by Firebase AppCheck) to verify that a genuine instance of your app is accessing your backend resources and that the device accessing it has not been tampered with.

### Why do we need it for mobile?

> As a public client of STS, our service is susceptible to malicious attacks from unauthorized clients. We have therefore implemented app integrity checks when the `/token` endpoint is called to ensure that:

Reviewer: I think this is slightly the wrong way around:
I think is STS supports public clients and because the client is public (i.e. not running on a server we control) we need a way to ensure that the client calling the service is a genuine (signed) instance of our app.

Author: I'm not sure I see a difference between the two 😭
What exactly is the client in this situation? Is it the app?

> As a public client of STS, our service is susceptible to malicious attacks from unauthorized clients. We have therefore implemented app integrity checks when the `/token` endpoint is called to ensure that:

Reviewer: ```suggestion
As a public client of STS, our service is susceptible to malicious attacks from unauthorized clients.

We have therefore implemented app integrity checks when the `/token` endpoint is called to ensure that the client is
```

> As a public client of STS, our mobile app is susceptible to malicious attacks from unauthorized clients. We have therefore implemented app integrity checks when the `/token` endpoint is called to ensure that:

Reviewer 2: It's the "service" which is susceptible to attack rather than the mobile app 🙂

1. genuine: the client is not being impersonated
2. unmodified: the code we have deployed has not been changed by the device

This therefore adds a layer of security to our authorization flow.

> 

Reviewer 3: @Magdaline-A Please may we amend this part to:

We have implemented app integrity checks to ensure that the mobile app is:

1. Genuine: the client is not being impersonated
2. Unmodified: The code we have deployed has not been changed by the device

> 1. the token is being issued to the genuine instance of the client requesting it, i.e. the client is not being impersonated

Reviewer: ```suggestion
1. genuine: the client is not being impersonated
2. unmodified: the code we have deployed has not been changed by the device
```

> 1. the access token is being issued to the genuine instance of the client requesting it, i.e. the client is not being impersonated

Reviewer 2: I would just say "tokens" rather than specifically the access token here

> 2. that the authorization code has not been intercepted ([PKCE](https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-pkce)).

Reviewer: I'm not sure PKCE is relevant to this package.

## Testing

### Testing locally

> Using the AppCheckDebugProviderFactory type when the build configuration is in Debug  allows us to pass a known debug token as an environment variable to Firebase App Check.

Reviewer: ```suggestion
Using the `AppCheckDebugProviderFactory` type when the build configuration is in Debug allows us to pass a known debug token as an environment variable to Firebase App Check.
```

This can be done by editing the scheme on xcode and passing an environment variable with the name FIRAAppCheckDebugToken.

> The value should be saved in an encrypted secrets file outside of the repository and passed in by including the file in your build config as shown below:

Reviewer 2: The file itself should be encrypted but should come from your ["CI system's secure key store"](https://firebase.google.com/docs/app-check/ios/debug-provider)
